kpops/utils/gen_schema.py

```diff
@@ -0,0 +1,137 @@
+
+
+class SchemaScope(str, Enum):
+    pipeline = "pipeline"
+    config = "config"
+
+
+original_field_schema = schema.field_schema
+
+
+# adapted from https://github.com/tiangolo/fastapi/issues/1378#issuecomment-764966955
+def field_schema(field: ModelField, **kwargs: Any) -> Any:
+    if field.field_info.extra.get("hidden_from_schema"):
+        raise SkipField(f"{field.name} field is being hidden")
+    else:
+        return original_field_schema(field, **kwargs)
+
+
+schema.field_schema = field_schema
+
+log = logging.getLogger("")
+
+ModelField
+
+
+def is_valid_component(
+    used_fields: list[str], component: type[PipelineComponent]
+) -> bool:
+    """
+    Check whether a PipelineComponent subclass has a valid definition for the schema generation.
+
+    :param used_fields: types defined so far
+    :type used_fields: list[str]
+    :param component: component type to be validated
+    :type component: type[PipelineComponent]
+    :return: Whether component is valid for schema generation
+    :rtype: bool
+    """
+    component_name: str = component.__name__
+    # PipelineComponent always has a field 'type'
+    type: str = component.__fields__.get("type").field_info.default  # type: ignore[union-attr]
+    if not component.__fields__.get("schema_type"):
+        log.warning(f"SKIPPED {component_name}, schema_type is not defined")
+        return False
+    schema_type: ModelField = component.__fields__.get("schema_type")  # type: ignore[assignment]
+    schema_type_default = schema_type.field_info.default
+    if (
+        get_origin(schema_type.type_) is not Literal
+        or len(get_args(schema_type.type_)) != 1
+    ):
+        log.warning(
+            f"SKIPPED {component_name}, schema_type must be a Literal with 1 argument of type str."
+        )
+        return False
+    elif get_args(schema_type.type_)[0] != schema_type_default:
+        log.warning(
+            f"SKIPPED {component_name}, schema_type default value must match Literal arg"
+        )
+        return False
+    elif schema_type_default != type:
+        log.warning(f"SKIPPED {component_name}, schema_type != type.")
+        return False
+    elif schema_type_default in used_fields:
+        log.warning(f"SKIPPED {component_name}, schema_type must be unique.")
+        return False
+    else:
+        used_fields.append(schema_type_default)
+        return True
+
+
+def gen_pipeline_schema(
+    components_module: str | None = None,
+) -> None:
+    """
+    Generate a json schema from the models of pipeline components.
+
+    :param components_module: Python module. Only the classes that inherit from PipelineComponent will be considered.
+    :type components_module: str or None
+    :rtype: None
+    """
+
+    components = tuple(_find_classes("kpops.components", PipelineComponent))
+    if components_module:
+        # record components types used so far
+        used_fields: list[str] = [
```

Comment:
what's your opinion on using `set` instead of `list`? it's more efficient to do lookups against it

Reply:
`set` seems like a good idea :)

Reply:
done